src/python/grpcio_testing/grpc_testing/__init__.py

```diff
@@ -0,0 +1,441 @@
+        metadata to the RPC (or has provided one or more response messages or
+        has terminated the RPC, either of which will cause gRPC Python to
+        synthesize initial metadata for the RPC).
+
+        Returns:
+          The initial metadata for the RPC.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def add_requests_as_messages(self, requests):
+        """Accepts requests to be passed to the system under test.
+
+        Args:
+          requests: An iterable of deserialized request messages to be passed
+            to the system under test.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def take_response_as_message(self):
+        """Draws one of the responses added to the RPC by the system under test.
+
+        Successive calls to this method return responses in the same order in
+        which the system under test added them to the RPC.
+
+        Returns:
+          A deserialized response message added to the RPC by the system under
+            test.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def no_more_requests(self):
+        """Indicates the end of the RPC's request stream."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def cancel(self):
+        """Cancel's the RPC."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def terminate(self):
+        """Blocks until the system under test has terminated the RPC."""
+        raise NotImplementedError()
+
+
+class ServerFixture(six.with_metaclass(abc.ABCMeta)):
+    """Fixture for a server in which a system under test services RPCs."""
+
+    @abc.abstractmethod
+    def invoke_rpc_by_service_and_method_names(self, service_name, method_name,
+                                               invocation_metadata, requests,
+                                               no_more_requests, timeout):
+        """Invokes an RPC to be serviced by the system under test.
+
+        Args:
+          service_name: An RPC service name (such as "my_package.MyService").
+          method_name: An RPC method name (such as "MyMethod").
+          invocation_metadata: The metadata supplied at RPC-invocation-time by
+            the invoker of the RPC.
+          requests: An iterable of request messages with which to start the RPC.
+            These are just a start; they need not be all requests that will ever
+            be added to the RPC.
+          no_more_requests: True if all requests are known and passed in this
+            call and the RPC's request stream should be closed; False if test
+            code may add more requests later or wishes to hold the request
+            stream open until a later time.
+          timeout: A duration of time in seconds for the RPC.
+        """
+        raise NotImplementedError()
+
+
+class Connection(six.with_metaclass(abc.ABCMeta)):
```

Comment:
What is the benefit to testing a system that is both a server and a client?  Should we encourage users to test server and client components in isolation?